src/librustc/traits/error_reporting.rs

```diff
@@ -835,42 +854,51 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
         }
     }
 
-    fn need_type_info(&self, obligation: &PredicateObligation<'tcx>, ty: Ty<'tcx>) {
+    pub fn need_type_info(&self, body_id: hir::BodyId, span: Span, ty: Ty<'tcx>) {
```

Comment:
Nit for future PR: feels like this function (and its associated visitor etc) would be ripe for promotion into a sub-module `error_reporting/need_type_info.rs`